source/s3_default_buffer_pool.c

```diff
@@ -0,0 +1,612 @@
+        AWS_LOGF_WARN(
+            AWS_LS_S3_CLIENT,
+            "Part size specified on the client is too large for automatic buffer reuse. "
+            "Consider specifying a smaller part size to improve performance and memory utilization");
+        chunk_size = 0;
+    }
+
+    struct aws_s3_default_buffer_pool *buffer_pool =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_default_buffer_pool));
+
+    AWS_FATAL_ASSERT(buffer_pool != NULL);
+
+    buffer_pool->base_allocator = allocator;
+    buffer_pool->chunk_size = chunk_size;
+    buffer_pool->block_size = s_chunks_per_block * chunk_size;
+    /* Somewhat arbitrary number.
+     * Tries to balance between how many allocations use buffer and buffer space
+     * being wasted. */
+    buffer_pool->primary_size_cutoff = chunk_size * 4;
+    buffer_pool->mem_limit = adjusted_mem_lim;
+    int mutex_error = aws_mutex_init(&buffer_pool->mutex);
+    AWS_FATAL_ASSERT(mutex_error == AWS_OP_SUCCESS);
+
+    aws_array_list_init_dynamic(
+        &buffer_pool->blocks, allocator, s_block_list_initial_capacity, sizeof(struct s3_buffer_pool_block));
+
+    aws_linked_list_init(&buffer_pool->pending_reserves);
+
+    struct aws_s3_buffer_pool *pool = aws_mem_calloc(allocator, 1, sizeof(struct aws_s3_buffer_pool));
+    pool->impl = buffer_pool;
+    pool->vtable = &s_default_pool_vtable;
+    aws_ref_count_init(&pool->ref_count, pool, (aws_simple_completion_callback *)aws_s3_default_buffer_pool_destroy);
+
+    return pool;
+}
+
+void aws_s3_default_buffer_pool_destroy(struct aws_s3_buffer_pool *buffer_pool_wrapper) {
+    if (buffer_pool_wrapper == NULL) {
+        return;
+    }
+    AWS_FATAL_ASSERT(buffer_pool_wrapper->impl);
+
+    struct aws_s3_default_buffer_pool *buffer_pool = buffer_pool_wrapper->impl;
+    aws_mem_release(buffer_pool->base_allocator, buffer_pool_wrapper);
+
+    for (size_t i = 0; i < aws_array_list_length(&buffer_pool->blocks); ++i) {
+        struct s3_buffer_pool_block *block;
+        aws_array_list_get_at_ptr(&buffer_pool->blocks, (void **)&block, i);
+
+        AWS_FATAL_ASSERT(block->alloc_bit_mask == 0 && "Allocator still has outstanding blocks");
+        aws_mem_release(buffer_pool->base_allocator, block->block_ptr);
+    }
+
+    aws_array_list_clean_up(&buffer_pool->blocks);
+
+    for (struct aws_linked_list_node *node = aws_linked_list_begin(&buffer_pool->pending_reserves);
+         node != aws_linked_list_end(&buffer_pool->pending_reserves);
+         node = aws_linked_list_next(node)) {
+        struct s3_pending_reserve *pending = AWS_CONTAINER_OF(node, struct s3_pending_reserve, node);
+        aws_future_s3_buffer_ticket_release(pending->ticket_future);
+        aws_linked_list_remove(node);
+    }
+
+    aws_mutex_clean_up(&buffer_pool->mutex);
+    struct aws_allocator *base = buffer_pool->base_allocator;
+    aws_mem_release(base, buffer_pool);
+}
+
+void s_buffer_pool_trim_synced(struct aws_s3_default_buffer_pool *buffer_pool) {
```

Comment:
The function name `s_try_reserve` doesn't clearly indicate what it does. A better name might be `s_attempt_buffer_reservation` or `s_try_allocate_buffer_ticket`.